### PR TITLE
Resistant event subs

### LINKED ISSUE #40: Resistant Event Sub
**Location**: Package `[backend/eth/glue]`, part of #19

**Motivation**: The Resistant Event Sub tracks the state of an event in the case of a chain reorganization.  

**Design**: Analogous to #39 the design is very simple. It wraps an `EventSub` and accepts a `finalityDepth`. The necessary `ChainReader` is provided by the `ContractBackend`.  

It tracks the status of an Event by subscribing to new block chain headers and then checking all events for finality.  
Once an Event is final, it is emitted on the `sink` channel that the user passed to `Read`.

uml
[uml.txt](https://github.com/hyperledger-labs/go-perun/files/6280360/uml.txt)

**Testing**: So far i am testing that final Events arrive even after reorgs. Still needs to be slightly extended.

### PR DESCRIPTION
**Pull request by ggwpez**
Adds the `ResistantEventSub` type.  
Also introduced a convenience function `Subscribe(ctx context.Context, cr ethereum.ChainReader, contract *bind.BoundContract, eFact EventFactory, startBlockOffset, confirmations uint64) (*ResistantEventSub, error)` which makes it easier to create a resistant sub.  
I am not happy with the arguments of `Subscribe` yet, since it are so many.

Closes #40 
Draft until #105 is closed

### CODE REVIEW

File: backend/ethereum/subscription/eventsub.go [hunk 1]
```diff
@@ -199,25 +205,25 @@ func (s *EventSub) readFuture(ctx context.Context, sink chan<- *Event) error {
...
-// Should be called exactly once and panics otherwise.
-// Must not be called if the construction function returned with an error.
+// Can be called more than once, but should be called at least once.
+// Is thread safe.
 func (s *EventSub) Close() {
-	close(s.closed)
-	s.watchSub.Unsubscribe()
-	s.filterSub.Unsubscribe()
+	if err := s.closer.Close(); !pkgsync.IsAlreadyClosedError(err) {
+		log.WithError(err).Error("could not close EventSub")
```
Comment (ggwpez) [inline backend/ethereum/subscription/eventsub.go:227]: Regarding the error ignore here:  
I changed it to return the error but the change would propagate up to the usage of the `RegisteredSub` so i would rather do this in a separate issue.

File: backend/ethereum/subscription/eventsub.go [hunk 2]
```diff
@@ -24,14 +24,16 @@ import (
 	"github.com/ethereum/go-ethereum/core/types"
```
Comment (ggwpez) [inline backend/ethereum/subscription/eventsub.go:24]: Will fix the commit message after the first review.

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		}
+		log.Debug("Waiting for confirm")
+		OneEvent(require, sub)
+		sub.Close()
+		s.SB.Commit() // Make sure the next EventSub does not get some old block.
+	}
+}
+
+// TestResistantEventSub_FinalityDepth tests that a TX is confirmed exactly
+// after `finalityDepth` blocks even when reorgs occur.
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:71]:
```suggestion
// TestResistantEventSub_Reorg tests that a TX is confirmed exactly
// after `finalityDepth` blocks even when reorgs occur.
```

Also: Here you call it `finalityDepth`, above you call it confirmations. Either one or the other should be consistently used. I prefer `finalityDepth`.

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	require := require.New(t)
+	s := test.NewTokenSetup(ctx, t, rng)
+
+	for i := 0; i < 3; i++ {
+		finality := rng.Int31n(20) + 2
+		sub, err := subscription.Subscribe(ctx, s.CB, s.Contract, event, 0, uint64(finality))
+		require.NoError(err)
+
+		for i := 0; i < 20; i++ {
+			s.SB.Commit()
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:85]: 20 empty blocks? What is this supposed to simulate?

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	rng := pkgtest.Prng(t)
+	require := require.New(t)
+	s := test.NewTokenSetup(ctx, t, rng)
+
+	for i := 0; i < 3; i++ {
+		finality := rng.Int31n(20) + 2
+		sub, err := subscription.Subscribe(ctx, s.CB, s.Contract, event, 0, uint64(finality))
+		require.NoError(err)
+
+		for i := 0; i < 20; i++ {
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:84]: If this `20` here and the one above in line 80 refer to the same thing, you should use a const.

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+			s.SB.Commit()
+		}
+		// Send and Confirm the TX
+		s.IncAllowance(ctx)
+		// Reorg until the block hight hits `finality` blocks.
+		for h := int64(0); h < int64(finality-1); {
+			d := rng.Int63n(10) + 1
+			l := rng.Int63n(10) + 1
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:92]: If the value 10 relates to the value 20 above in some way, I suggest to use a constant to ensure consistency in case one of the values is changed.

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+			log.Debugf("[h=%d] Reorg with depth: %d, length: %d", h, d, l)
+			s.SB.Reorg(ctx, uint64(d), func(txs []types.Transactions) []types.Transactions {
+				return append(txs, make([]types.Transactions, int(d+l)-len(txs))...)
+			})
+			h += l
+		}
+		log.Debug("done")
+		// Verify that the event arrived.
+		OneEvent(require, sub)
+		sub.Close()
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:103]: Above, you add another `s.SB.Commit()` here. Looks like we don't need it? Or we should also add it here?

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	for i := 0; i < n; i++ {
+		require.NotNil(<-sink)
+	}
+
+	select {
+	case event := <-sink:
+		require.Nil(event)
+	default:
+	}
+}
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:143]: What do you think about adding a test for the case that an event is removed by a reorg and should therefore not be emitted?

File: backend/ethereum/subscription/resistanteventsub.go [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+	if diff.Sign() > 0 {
+		for _, event := range s.events {
+			if s.isFinal(event) {
+				sink <- event
+				delete(s.events, event.Log.TxHash)
+			}
+		}
+	} else {
+		log.Debugf("Block number did not increase. Increment: %v", diff)
+	}
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub.go:222]: It is possible to just remove the check, the check just optimizes it slightly by only checking on increased block numbers.

File: backend/ethereum/subscription/eventsub.go (hunk 1 above)
Comment (ggwpez) [reply backend/ethereum/subscription/eventsub.go:227]: Move to #162

File: backend/ethereum/subscription/resistanteventsub.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+		lastBlockNum *big.Int
+		heads        chan *types.Header
+		headSub      ethereum.Subscription
+		events       map[common.Hash]*Event
+	}
+)
+
+// NewResistantEventSub creates a new `ResistantEventSub` from the given
+// `EventSub`. Closes the passed `EventSub` when done.
+// `confirmations` defines in how many blocks the events needs to be included.
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub.go:67]: Yes will rename it, was annoyed from that as well.

File: backend/ethereum/subscription/resistanteventsub.go [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+		case <-s.closer.Closed():
+			return nil
+		}
+	}
+}
+
+// ReadPast reads all past events into `sink`.
+// Can be aborted by cancelling `ctx` or `Close()`.
+// All events can be considered final.
+func (s *ResistantEventSub) ReadPast(_ctx context.Context, sink chan<- *Event) error {
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub.go:147]: Probably needed in the next MR, will check.

File: backend/ethereum/subscription/resistanteventsub.go [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+				sink <- event
+				delete(s.events, event.Log.TxHash)
+			}
+		}
+	} else {
+		log.Debugf("Block number did not increase. Increment: %v", diff)
+	}
+}
+
+func (s *ResistantEventSub) isFinal(event *Event) bool {
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub.go:225]: Yes, things like that where the header and the event sub race each other can happen.  
https://github.com/hyperledger-labs/go-perun/issues/157 contains a similar case.

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 4 above)
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:85]: The blocks here are needed so that the following `Reorg` does not reorg blocks from the previous loop iteration.  
I think i will just remove the encompassing loop and only test it once.

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 5 above)
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:84]: I added a `maxFinality` constant.

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 6 above)
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:92]: Done

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 7 above)
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:103]: I remove the loops, so it is superfluous now.

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 8 above)
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:143]: Ok makes sense, analogous to the TX finality test.

Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:143]: Wow… adding such a test fails, since indeed, the header subscription seems to be much faster than the event subscription.

File: backend/ethereum/subscription/resistanteventsub.go [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+		events:        make(map[common.Hash]*Event),
+	}
+	ret.closer.OnCloseAlways(func() {
+		headSub.Unsubscribe()
+		sub.Close()
+	})
+	return ret, nil
+}
+
+// Read reads all past and future events into `sink`.
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub.go:107]: We need all past event for example in the case of the `Funder`, which sums up all `Deposited` events.  
The `AdjSub` makes it easy to access the last `ChannelUpdate` event, where it does not really matter if you know the previous events or not.

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 3 above)
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:71]: I renamed it everywhere.

File: backend/ethereum/subscription/resistanteventsub.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub.go:147]: I added `TestResistantEventSub_ReadPast`, besides that it will be used in #117, .

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+	require := require.New(t)
+	s := test.NewTokenSetup(ctx, t, rng)
+
+	finality := rng.Int31n(maxFinality-2) + 2
+	sub, err := subscription.Subscribe(ctx, s.CB, s.Contract, event, 0, uint64(finality))
+	require.NoError(err)
+	defer sub.Close()
+
+	// Send and Confirm the TX
+	s.ConfirmTx(s.IncAllowance(ctx), true)
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:175]: I have trouble following. Is it correct, that `ConfirmTx` waits until `TxFinalityDepth = 6` blocks until it returns? How is that the case here, if we do not progress the chain?

Update:
OK, checked in more detail. Realized that you set `TxFinalityDepth = 1` in the init function for this test. It doesn't seem like a good idea that we specify `TxFinalityDepth` as a global variable. It's hard to follow which value it currently has and there is also no protection against changing it on-the-fly, which may result in undefined behavior. Anyways, this is out of scope here. I will create an issue for that.

Comment (matthiasgeihs) [reply backend/ethereum/subscription/resistanteventsub_test.go:175]: Do we need to call `ConfirmTx` here at all?

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+		return make([]types.Transactions, 2)
+	})
+
+	time.Sleep(1 * time.Second) // give geth time to update
+	NoEvent(require, sub)
+	// Verify that the event never arrives.
+	for i := 0; i < maxFinality; i++ {
+		s.SB.Commit()
+	}
+	time.Sleep(1 * time.Second) // give geth time to update
```
Comment (matthiasgeihs) [inline backend/ethereum/subscription/resistanteventsub_test.go:189]:
```suggestion
	time.Sleep(1 * time.Second) // give the event subscription time to catch up
```

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 8 above)
Comment (matthiasgeihs) [reply backend/ethereum/subscription/resistanteventsub_test.go:229]: Looks like to increased that to a 1second wait.

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -64,4 +64,4 @@ jobs:
         uses: actions/checkout@v2
 
       - name: Unit Race Tests
-        run: go test -timeout 120s -race -gcflags=all=-d=checkptr=0 ./...
+        run: go test -timeout 240s -race -gcflags=all=-d=checkptr=0 ./...
```
Comment (matthiasgeihs) [inline .github/workflows/ci.yml:67]: Not nice :( Does that mean our CI runs twice as long now?

Update:
I compared the running time of a CI run before and after the update. They are roughly equal. I don't think we need to bump the timeout, do we?

File: backend/ethereum/subscription/resistanteventsub_test.go (hunk 14 above)
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go:175]: Right, i just added that yesterday for debugging, removed.

File: .github/workflows/ci.yml (hunk 16 above)
Comment (ggwpez) [reply .github/workflows/ci.yml:67]: The i dont know why it time-outed once… I will try it without again.

File: backend/ethereum/subscription/resistanteventsub_test.go [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+	// Send and Confirm the TX
+	s.ConfirmTx(s.IncAllowance(ctx), true)
+	NoEvent(require, sub)
+
+	// Go back one block and remove the TX with a reorg.
+	s.SB.Reorg(ctx, 1, func(txs []types.Transactions) []types.Transactions {
+		return make([]types.Transactions, 2)
+	})
+
+	time.Sleep(1 * time.Second) // give geth time to update
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/subscription/resistanteventsub_test.go]: I removed it.

**Review by matthiasgeihs — APPROVED** [review]
